crates/circuit/src/dag_circuit.rs

```diff
@@ -6918,34 +6890,36 @@ impl DAGCircuit {
         }
         let build_var_mapper =
             |cregs: &RegisterData<ClassicalRegister>| -> PyResult<PyVariableMapper> {
-                let edge_map = if qubit_map.is_empty() && clbit_map.is_empty() {
-                    // try to ido a 1-1 mapping in order
-                    let out_dict = PyDict::new(py);
-                    for (a, b) in identity_qubit_map.iter() {
-                        out_dict.set_item(a.into_py_any(py)?, b.into_pyobject(py)?)?;
-                    }
-                    for (a, b) in identity_clbit_map.iter() {
-                        out_dict.set_item(a.into_py_any(py)?, b.into_pyobject(py)?)?;
-                    }
-                    out_dict
-                } else {
-                    let out_dict = PyDict::new(py);
-                    for (a, b) in qubit_map.iter() {
-                        out_dict.set_item(a.into_py_any(py)?, b.into_pyobject(py)?)?;
-                    }
-                    for (a, b) in clbit_map.iter() {
-                        out_dict.set_item(a.into_py_any(py)?, b.into_pyobject(py)?)?;
-                    }
-                    out_dict
-                };
+                Python::with_gil(|py| {
```

Comment:
I can live with this because it _should_ be temporary (though #14300 is only for `substitute_node_with_dag` and we should track `compose` too), but in general I think it's a mistake to hide the `with_gil` inside the lambda.  Whenever we _must_ acquire the lock when a function is called, it's asking for deadlocks / surprising behaviour / paper-cut slowdowns by acquiring/reacquiring it when we hide it away from the caller.  The caller still has to acquire the lock, it just doesn't _know_ it has to acquire the lock.

Also, if nothing else, making the caller do it removes a level of indentation.